src/changelog/prepareChangelogPaths.ts

```diff
@@ -0,0 +1,142 @@
+  md?: string;
+  /** Absolute path to changelog json file (new or existing) */
+  json?: string;
+  /** Suffix being used in files, if any */
+  suffix?: string;
+}
+
+/**
+ * Get the paths to the changelog files. Also handles conversion between `changelog.uniqueFilenames`
+ * being true and false/unset (moving the files if needed).
+ *
+ * @returns object with each changelog path, or undefined if that changelog shouldn't be written.
+ */
+export function prepareChangelogPaths(params: {
+  options: Pick<BeachballOptions, 'changelog' | 'generateChangelog'>;
+  /** cwd where changelogs are located */
+  changelogAbsDir: string;
+}): ChangelogPaths {
+  const { options, changelogAbsDir: cwd } = params;
+  const { changelog, generateChangelog } = options;
+  const { uniqueFilenames } = changelog || {};
+
+  const paths: ChangelogPaths = {};
+  let existingSuffixedPaths: ChangelogPaths | undefined;
+
+  if (!generateChangelog) {
+    return paths;
+  }
+
+  for (const ext of ['md', 'json'] as const) {
+    if (generateChangelog !== true && generateChangelog !== ext) {
+      continue; // not generating this file type
+    }
+
+    const defaultPath = path.join(cwd, `CHANGELOG.${ext}`);
+
+    if (uniqueFilenames) {
+      // Filenames should use a suffix. Start by getting any existing suffixed changelogs...
+      existingSuffixedPaths ??= _getExistingSuffixedChangelogs(cwd);
+      // Use the existing suffix or generate a new one
+      paths.suffix ??= existingSuffixedPaths.suffix ?? uuid().slice(0, 8);
+
+      if (existingSuffixedPaths[ext]) {
+        // Suffixed file already exists, so use it
+        paths[ext] = existingSuffixedPaths[ext];
+      } else {
+        // New suffixed file
+        paths[ext] = path.join(cwd, `CHANGELOG-${paths.suffix}.${ext}`);
+        // Move any existing non-suffixed file to the new suffixed path
+        moveIfNeeded({ oldPath: defaultPath, newPath: paths[ext]!, hadSuffix: false });
+      }
+    } else {
+      // Filenames should not use a suffix
+      paths[ext] = defaultPath;
+
+      if (!fs.existsSync(defaultPath)) {
+        // If the default file doesn't exist, check for a suffixed file to move back
+        existingSuffixedPaths ??= _getExistingSuffixedChangelogs(cwd);
+        if (existingSuffixedPaths[ext]) {
+          moveIfNeeded({ oldPath: existingSuffixedPaths[ext]!, newPath: defaultPath, hadSuffix: true });
+        }
+      }
+    }
+  }
+
+  return paths;
+}
+
+/**
+ * Get paths to existing `CHANGELOG-{suffix}.md` and `CHANGELOG-{suffix}.json` files.
+ * Exported for testing only.
+ */
+export function _getExistingSuffixedChangelogs(cwd: string): ChangelogPaths {
+  let suffixedFiles: { file: string; suffix: string; date?: number }[];
+  try {
+    suffixedFiles = fs
+      .readdirSync(cwd)
+      .filter(file => /^CHANGELOG-[a-f\d]{8,}\.(json|md)$/.test(file))
+      .map(file => ({ file, suffix: file.match(/[a-f\d]{8,}/)![0] }));
```

Comment:
This regex really depends on the fact that CHANGELOG- and ".json" or ".md" have enough characters that do not match the hexadecimal representation. To me, it seems like we should strip the file extension and take the last eight characters after that instead of relying on regex magic. (But I have a bias against regexes when string manipulation is straightforward.)

Reply:
> This regex really depends on the fact that CHANGELOG- and ".json" or ".md" have enough characters that do not match the hexadecimal representation. 

Not sure what you mean by this? The regex is very specific: `^` start of string, `CHANGELOG-`, 8+ hex characters, `.md` or `.json`, `$` end of string.

Reply:
Perhaps I'm misunderstanding things, but I see two regexes and the first filters the filenames based on the match that includes `^` and `$` but the second has a group match that doesn't have those protections. Is this `file.match()` somehow reusing the previous regex information?

So what I was trying to say is that the regex `/[a-f\d]{8,}/` will find a match, and relies on the fact that a string matching `/^CHANGELOG-[a-f\d]{8,}\.(json|md)$/` has only one possible substring to match that hex string. (This would not be the case for a string `deadbeef-12345678.json`)

Reply:
Sorry about that, for some reason I initially read your comment as pointing to the next line up (filter). Then after posting I realized you were actually talking about the map line, which makes more sense.